src/persist/src/operators/stream.rs

```diff
@@ -721,6 +694,77 @@ mod tests {
         Ok(())
     }
 
+    /// Test to make sure we handle closing the seal operator correctly and don't
+    /// incorrectly process any seal futures after the operator has been closed.
+    #[test]
+    fn seal_close() -> Result<(), Error> {
+        ore::test::init_logging_default("trace");
+        let mut registry = MemRegistry::new();
+        let mut unreliable = UnreliableHandle::default();
+        let p = registry.runtime_unreliable(unreliable.clone())?;
+
+        timely::execute_directly(move |worker| {
+            let (mut input, mut dummy, probe) = worker.dataflow(|scope| {
+                let (write, _read) = p.create_or_load::<(), ()>("primary");
+                let mut input = Handle::new();
+                let stream = input.to_stream(scope);
+                // We need to create a dummy stream to force the dataflow to stay around
+                // even after the actual input has been closed.
+                let mut dummy = Handle::new();
+                let dummy_stream = dummy.to_stream(scope);
+
+                let sealed_stream = stream.seal("test", write);
+
+                let stream = dummy_stream.concat(&sealed_stream);
+                let probe = stream.probe();
+
+                (input, dummy, probe)
+            });
+
+            // Send data here mostly to avoid having to dictate types to the
+            // compiler.
+            input.send((((), ()), 0, 1));
+            dummy.send((((), ()), 0, 1));
+
+            dummy.advance_to(1);
+            unreliable.make_unavailable();
+
+            // Advance the frontier while persist is unavailable in order to force
+            // the seal operator into a retry loop with this seal operation.
+            input.advance_to(1);
+
+            // We intentionally take a thousand steps here because taking fewer steps
+            // makes the failure much less likely. We can't simply wait for frontier
+            // advancement because persistence is unavailable, so sealing will fail.
+            for _ in 0..1_000 {
```

Comment:
how long does this test take?

Reply:
1/2 a sec

Reply:
every non-s3 test in src/persist (79 of them) totals `1.82s` on my machine right now, so that's probably fine but unfortunate. does aljoscha's commit help this?

Reply:
Sorry. I gave you the number to do `time cargo test -- seal_done`, which I now realize is incredibly misleading because cargo test takes a lot of time to set up.

On my laptop, all tests take 2.60 seconds (insert m1 mac joke here), with the expensive version of the test they take 2.70, and with aljoscha's change they take ~2.62 seconds, so the overall effect was not that bad.